File: core/src/test/java/com/jlpka/langidentify/WordSegmenterTest.java
package com.jlpka.langidentify;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

/**
 * Tests for WordSegmenter: word splitting, lowercasing, apostrophe handling, digit filtering,
 * alphabet boundary detection, and CJ run routing.
 */
class WordSegmenterTest {

  // A model with English + French (both LATIN) — no CJ languages, no CJClassifier.
  static Model latinModel;

  // A model with English + Japanese (LATIN + HAN + JA_KANA).
  // Only one CJ language, so no CJClassifier instantiation.
  static Model cjModel;

  /** Captured word emission from WordConsumer. */
  static class Word {
    final String text;
    final int alphaIdx;

    Word(String text, int alphaIdx) {
      this.text = text;
      this.alphaIdx = alphaIdx;
    }
  }

  /** Collects words and CJ runs into lists. */
  static class Collector {
    final List<Word> words = new ArrayList<>();
    final List<Word> cjRuns = new ArrayList<>();
  }

  @BeforeAll
  static void setup() throws IOException {
    // Latin-only model: en + fr share LATIN alphabet
    latinModel =
        new Model(
            new NgramTable[] {new NgramTable.Builder().compact()},
            new Language[] {Language.ENGLISH, Language.FRENCH},
            -10.0,
            -10.0,
            1,
            null,
            null);

    // CJ model: en (LATIN) + ja (HAN, JA_KANA)
    // Only one CJ language, so needsCJ = false, no CJClassifier file needed
    cjModel =
        new Model(
            new NgramTable[] {new NgramTable.Builder().compact()},
            new Language[] {Language.ENGLISH, Language.JAPANESE},
            -10.0,
            -10.0,
            1,
            null,
            null);
  }

  private WordSegmenter makeSegmenter(Model model, boolean withCJ, Collector col) {
    return new WordSegmenter(
        model,
        (wordBuf, wordLen, alphaIdx) ->
            col.words.add(new Word(new String(wordBuf, 0, wordLen), alphaIdx)),
        withCJ
            ? (wordBuf, wordLen, alphaIdx) ->
                col.cjRuns.add(new Word(new String(wordBuf, 0, wordLen), alphaIdx))
            : null);
  }

  private List<Word> segmentWords(String text) {
    var col = new Collector();
    var seg = makeSegmenter(latinModel, false, col);
    seg.segment(text);
    return col.words;
  }

  // ========================================================================
  // Basic word splitting
  // ========================================================================

  @Test
  void emptyString() {
    assertEquals(List.of(), segmentWords(""));
  }

  @Test
  void singleWord() {
    var words = segmentWords("hello");
    assertEquals(1, words.size());
    assertEquals("hello", words.get(0).text);
  }

  @Test
  void multipleWords() {
    var words = segmentWords("hello world foo");
    assertEquals(3, words.size());
    assertEquals("hello", words.get(0).text);
    assertEquals("world", words.get(1).text);
    assertEquals("foo", words.get(2).text);
  }

  @Test
  void leadingAndTrailingSpaces() {
    var words = segmentWords("  hello  ");
    assertEquals(1, words.size());
    assertEquals("hello", words.get(0).text);
  }

  @Test
  void multipleSeparators() {
    var words = segmentWords("hello---world");
    assertEquals(2, words.size());
    assertEquals("hello", words.get(0).text);
    assertEquals("world", words.get(1).text);
  }

  // ========================================================================
  // Lowercasing
  // ========================================================================

  @Test
  void lowercasesWords() {
    var words = segmentWords("Hello WORLD FoO");
    assertEquals("hello", words.get(0).text);
    assertEquals("world", words.get(1).text);
    assertEquals("foo", words.get(2).text);
  }

  // ========================================================================
  // Apostrophe handling
  // ========================================================================

  @Test
  void apostropheSplitsWord() {
    var words = segmentWords("don't");
    assertEquals(1, words.size());
    assertEquals("don't", words.get(0).text);
  }

  @Test
  void smartApostrophe() {
    var words = segmentWords("don\u2019t");
    assertEquals(1, words.size());
    assertEquals("don't", words.get(0).text);
    words = segmentWords("don\u0092t");
    assertEquals(1, words.size());
    assertEquals("don't", words.get(0).text);
  }

  @Test
  void trailingApostropheStripped() {
    // "dogs'" -> apostrophe at end is stripped
    var words = segmentWords("dogs' ");
    assertEquals(1, words.size());
    assertEquals("dogs", words.get(0).text);
  }

  @Test
  void onlyApostropheAfterLetterDropped() {
    // Single letter followed by apostrophe at word boundary: "x'" -> trailing ' stripped -> "x"
    var words = segmentWords("x' y");
    assertEquals(2, words.size());
    assertEquals("x", words.get(0).text);
    assertEquals("y", words.get(1).text);
  }

  @Test
  void doubleApostropheNotAdded() {
    // "don''t" — second apostrophe should not be added (consecutive apostrophe check)
    var words = segmentWords("don''t");
    // The second ' is not a letter and not added, so it acts as a separator
    assertEquals(2, words.size());
    assertEquals("don", words.get(0).text);
    assertEquals("t", words.get(1).text);
  }

  // ========================================================================
  // Digit filtering (Latin only)
  // ========================================================================

  @Test
  void digitPoisonsLatinWord() {
    var words = segmentWords("abc123def");
    // The whole run is one Latin word with digits — should be filtered
    assertEquals(0, words.size());
  }

  @Test
  void digitBetweenLatinWords() {
    // Digits between words: "hello 42world" -> "hello" emitted, "42world" poisoned
    // Actually digits don't break the word since they're not alphabet chars,
    // but they set the HAS_DIGIT flag. "hello" is clean, then "42" sets digit flag,
    // then "world" is Latin with digit flag -> poisoned
    var words = segmentWords("hello 42world");
    assertEquals(1, words.size());
    assertEquals("hello", words.get(0).text);
  }

  @Test
  void pureDigitsNoEmission() {
    // Just digits, no alphabet chars — nothing emitted
    var words = segmentWords("12345");
    assertEquals(0, words.size());
  }

  // ========================================================================
  // Alphabet boundaries
  // ========================================================================

  @Test
  void alphabetBoundarySplitsWord() {
    // Latin followed immediately by Cyrillic (not in latin model, so those chars are ignored)
    // But with cjModel: Latin followed by HAN should split
    var col = new Collector();
    var seg = makeSegmenter(cjModel, false, col);
    // 'hello' (LATIN) + '\u4e00' (HAN)
    seg.segment("hello\u4e00");
    assertEquals(2, col.words.size());
    assertEquals("hello", col.words.get(0).text);
    assertEquals("\u4e00", col.words.get(1).text);
  }

  @Test
  void unknownAlphabetCharsIgnored() {
    // Characters not in the model's alphabet set are skipped
    var words = segmentWords("hello\u0410world"); // \u0410 = Cyrillic A, not in latinModel
    assertEquals(2, words.size());
    assertEquals("hello", words.get(0).text);
    assertEquals("world", words.get(1).text);
  }

  // ========================================================================
  // CJ consumer routing
  // ========================================================================

  @Test
  void cjUnifiedRoutedToCJConsumer() {
    var col = new Collector();
    var seg = makeSegmenter(cjModel, true, col);
    seg.segment("hello \u4e00\u4e8c\u4e09");

    // "hello" goes to wordConsumer
    assertEquals(1, col.words.size());
    assertEquals("hello", col.words.get(0).text);

    // HAN run goes to cjConsumer
    assertEquals(1, col.cjRuns.size());
    var run = col.cjRuns.get(0);
    assertEquals("\u4e00\u4e8c\u4e09", col.cjRuns.get(0).text);
  }

  @Test
  void jaKanaRoutedToCJConsumer() {
    var col = new Collector();
    var seg = makeSegmenter(cjModel, true, col);
    // Hiragana chars: \u3042 = あ, \u3044 = い, \u3046 = う
    seg.segment("hello \u3042\u3044\u3046");

    assertEquals(1, col.words.size());
    assertEquals("hello", col.words.get(0).text);

    // JA_KANA run goes to cjConsumer
    assertEquals(1, col.cjRuns.size());
    assertEquals("\u3042\u3044\u3046", col.cjRuns.get(0).text);
  }

  @Test
  void mixedKanaAndCJUnifiedSplitSeparately() {
    var col = new Collector();
    var seg = makeSegmenter(cjModel, true, col);
    // Kana then HAN: different alphabets, so two separate runs
    seg.segment("\u3042\u3044\u4e00\u4e8c");

    assertEquals(0, col.words.size());
    assertEquals(2, col.cjRuns.size());

    // First: JA_KANA
    var kana = col.cjRuns.get(0);
    assertEquals("\u3042\u3044", kana.text);
    // Second: HAN
    var cj = col.cjRuns.get(1);
    assertEquals("\u4e00\u4e8c", cj.text);
  }

  @Test
  void noCJConsumerMeansCJGoesToWordConsumer() {
    // When cjConsumer is null, CJ words go through wordConsumer
    var col = new Collector();
    var seg = makeSegmenter(cjModel, false, col);
    seg.segment("\u4e00\u4e8c");

    assertEquals(1, col.words.size());
    assertEquals(0, col.cjRuns.size());
    assertEquals("\u4e00\u4e8c", col.words.get(0).text);
  }

  @Test
  void cjRunOriginalTextOffset() {
    var col = new Collector();
    var seg = makeSegmenter(cjModel, true, col);
    String text = "abc \u4e00\u4e8c\u4e09 xyz";
    seg.segment(text);

    assertEquals(2, col.words.size()); // "abc", "xyz"
    assertEquals(1, col.cjRuns.size());
    assertEquals("\u4e00\u4e8c\u4e09", col.cjRuns.get(0).text);
  }

  // ========================================================================
  // Reuse across calls
  // ========================================================================

  @Test
  void segmenterReusableBetweenCalls() {
    var col = new Collector();
    var seg = makeSegmenter(latinModel, false, col);

    seg.segment("hello");
    assertEquals(1, col.words.size());
    assertEquals("hello", col.words.get(0).text);

    seg.segment("world foo");
    assertEquals(3, col.words.size());
    assertEquals("world", col.words.get(1).text);
    assertEquals("foo", col.words.get(2).text);
  }

  @Test
  void extractCJRuns() {
    var col = new Collector();
    var seg = makeSegmenter(cjModel, true, col);
    String blobLatin = "abcdefghij";
    String blobCJ = "那只敏捷的棕色狐狸跳过了那只懒惰的狗";
    StringBuilder text = new StringBuilder();
    text.append(blobLatin);
    text.append(blobCJ);
    for (int i = 0; i < WordSegmenter.MAX_WORD_LEN + 5; i += blobLatin.length()) {
      text.append(blobLatin);
    }
    for (int i = 0; i < WordSegmenter.MAX_CJ_WORD_LEN + 5; i += blobCJ.length()) {
      text.append(blobCJ);
    }
    seg.segment(text);
    assertEquals(2, col.words.size());
    assertEquals(3, col.cjRuns.size());
    assertEquals(blobLatin, col.words.get(0).text);
    assertEquals(blobCJ, col.cjRuns.get(0).text);
    assertEquals(WordSegmenter.MAX_WORD_LEN, col.words.get(1).text.length());
    assertEquals(WordSegmenter.MAX_CJ_WORD_LEN, col.cjRuns.get(1).text.length());
    String t = col.words.get(1).text;
    for (int i = 0; i < t.length(); i += blobLatin.length()) {
      String s = t.substring(i, Math.min(i + blobLatin.length(), t.length()));
      assertEquals(s, blobLatin.substring(0, s.length()));
    }
    t = col.cjRuns.get(1).text;
    for (int i = 0; i < t.length(); i += blobCJ.length()) {
      String s = t.substring(i, Math.min(i + blobCJ.length(), t.length()));
      assertEquals(s, blobCJ.substring(0, s.length()));
    }
  }

  // ========================================================================
  // Edge cases
  // ========================================================================

  @Test
  void onlySpaces() {
    assertEquals(List.of(), segmentWords("   "));
  }

  @Test
  void onlyPunctuation() {
    assertEquals(List.of(), segmentWords("---...!!!"));
  }

  @Test
  void singleCharWord() {
    var words = segmentWords("a b c");
    assertEquals(3, words.size());
    assertEquals("a", words.get(0).text);
    assertEquals("b", words.get(1).text);
    assertEquals("c", words.get(2).text);
  }

  @Test
  void digitDoesNotPoisonNonLatin() {
    // Digits adjacent to CJ chars should NOT poison the CJ word
    var col = new Collector();
    var seg = makeSegmenter(cjModel, false, col);
    seg.segment("123\u4e00\u4e8c");
    assertEquals(1, col.words.size());
    assertEquals("\u4e00\u4e8c", col.words.get(0).text);
  }
}

File: rust/langidentify/langidentify/tests/word_segmenter_test.rs
use langidentify::word_segmenter::{WordSegmenter, MAX_CJ_WORD_LEN, MAX_WORD_LEN};
use langidentify::{Language, Model};
use std::sync::Arc;

struct Word {
    text: String,
}

fn latin_model() -> Arc<Model> {
    let languages = Language::from_comma_separated("en,fr").unwrap();
    Arc::new(
        Model::load_embedded(
            langidentify_models_lite::resolve,
            &languages,
            -10.0,
            -10.0,
            0.0,
        )
        .unwrap(),
    )
}

fn cj_model() -> Arc<Model> {
    let languages = Language::from_comma_separated("en,ja").unwrap();
    Arc::new(
        Model::load_embedded(
            langidentify_models_lite::resolve,
            &languages,
            -10.0,
            -10.0,
            0.0,
        )
        .unwrap(),
    )
}

fn segment_words(text: &str) -> Vec<Word> {
    let model = latin_model();
    let mut seg = WordSegmenter::new();
    let mut words = Vec::new();
    seg.segment(text, &model, |buf, len, _aidx, _is_cj| {
        words.push(Word {
            text: buf[..len].iter().collect(),
        });
    });
    words
}

fn segment_with_model(text: &str, model: &Model) -> (Vec<Word>, Vec<Word>) {
    let mut seg = WordSegmenter::new();
    let mut words = Vec::new();
    let mut cj_runs = Vec::new();
    seg.segment(text, model, |buf, len, _aidx, is_cj| {
        let w = Word {
            text: buf[..len].iter().collect(),
        };
        if is_cj {
            cj_runs.push(w);
        } else {
            words.push(w);
        }
    });
    (words, cj_runs)
}

// ========================================================================
// Basic word splitting
// ========================================================================

#[test]
fn empty_string() {
    let words = segment_words("");
    assert_eq!(0, words.len());
}

#[test]
fn single_word() {
    let words = segment_words("hello");
    assert_eq!(1, words.len());
    assert_eq!("hello", words[0].text);
}

#[test]
fn multiple_words() {
    let words = segment_words("hello world foo");
    assert_eq!(3, words.len());
    assert_eq!("hello", words[0].text);
    assert_eq!("world", words[1].text);
    assert_eq!("foo", words[2].text);
}

#[test]
fn leading_and_trailing_spaces() {
    let words = segment_words("  hello  ");
    assert_eq!(1, words.len());
    assert_eq!("hello", words[0].text);
}

#[test]
fn multiple_separators() {
    let words = segment_words("hello---world");
    assert_eq!(2, words.len());
    assert_eq!("hello", words[0].text);
    assert_eq!("world", words[1].text);
}

// ========================================================================
// Lowercasing
// ========================================================================

#[test]
fn lowercases_words() {
    let words = segment_words("Hello WORLD FoO");
    assert_eq!("hello", words[0].text);
    assert_eq!("world", words[1].text);
    assert_eq!("foo", words[2].text);
}

// ========================================================================
// Apostrophe handling
// ========================================================================

#[test]
fn apostrophe_in_word() {
    let words = segment_words("don't");
    assert_eq!(1, words.len());
    assert_eq!("don't", words[0].text);
}

#[test]
fn smart_apostrophe() {
    // U+2019 right single quotation mark
    let words = segment_words("don\u{2019}t");
    assert_eq!(1, words.len());
    assert_eq!("don't", words[0].text);
    // U+0092 private use (Windows-1252 apostrophe)
    let words = segment_words("don\u{0092}t");
    assert_eq!(1, words.len());
    assert_eq!("don't", words[0].text);
}

#[test]
fn trailing_apostrophe_stripped() {
    let words = segment_words("dogs' ");
    assert_eq!(1, words.len());
    assert_eq!("dogs", words[0].text);
}

#[test]
fn only_apostrophe_after_letter_dropped() {
    let words = segment_words("x' y");
    assert_eq!(2, words.len());
    assert_eq!("x", words[0].text);
    assert_eq!("y", words[1].text);
}

#[test]
fn double_apostrophe_not_added() {
    let words = segment_words("don''t");
    assert_eq!(2, words.len());
    assert_eq!("don", words[0].text);
    assert_eq!("t", words[1].text);
}

// ========================================================================
// Digit filtering (Latin only)
// ========================================================================

#[test]
fn digit_poisons_latin_word() {
    let words = segment_words("abc123def");
    assert_eq!(0, words.len());
}

#[test]
fn digit_between_latin_words() {
    let words = segment_words("hello 42world");
    assert_eq!(1, words.len());
    assert_eq!("hello", words[0].text);
}

#[test]
fn pure_digits_no_emission() {
    let words = segment_words("12345");
    assert_eq!(0, words.len());
}

// ========================================================================
// Alphabet boundaries
// ========================================================================

#[test]
fn alphabet_boundary_splits_word() {
    let model = cj_model();
    let (words, _cj) = segment_with_model("hello\u{4e00}", &model);
    // "hello" is Latin, \u{4e00} is HAN — should be separate
    assert!(words.len() >= 1);
    assert_eq!("hello", words[0].text);
}

#[test]
fn unknown_alphabet_chars_ignored() {
    // Cyrillic А not in latin model
    let words = segment_words("hello\u{0410}world");
    assert_eq!(2, words.len());
    assert_eq!("hello", words[0].text);
    assert_eq!("world", words[1].text);
}

// ========================================================================
// CJ consumer routing
// ========================================================================

#[test]
fn cj_unified_routed_as_cj() {
    let model = cj_model();
    let (words, cj_runs) = segment_with_model("hello \u{4e00}\u{4e8c}\u{4e09}", &model);

    assert_eq!(1, words.len());
    assert_eq!("hello", words[0].text);

    assert_eq!(1, cj_runs.len());
    assert_eq!("\u{4e00}\u{4e8c}\u{4e09}", cj_runs[0].text);
}

#[test]
fn ja_kana_routed_as_cj() {
    let model = cj_model();
    let (words, cj_runs) = segment_with_model("hello \u{3042}\u{3044}\u{3046}", &model);

    assert_eq!(1, words.len());
    assert_eq!("hello", words[0].text);

    assert_eq!(1, cj_runs.len());
    assert_eq!("\u{3042}\u{3044}\u{3046}", cj_runs[0].text);
}

#[test]
fn mixed_kana_and_han_split_separately() {
    let model = cj_model();
    let (_words, cj_runs) = segment_with_model("\u{3042}\u{3044}\u{4e00}\u{4e8c}", &model);

    assert_eq!(2, cj_runs.len());
    assert_eq!("\u{3042}\u{3044}", cj_runs[0].text);
    assert_eq!("\u{4e00}\u{4e8c}", cj_runs[1].text);
}

#[test]
fn cj_run_with_latin() {
    let model = cj_model();
    let (words, cj_runs) = segment_with_model("abc \u{4e00}\u{4e8c}\u{4e09} xyz", &model);

    assert_eq!(2, words.len()); // "abc", "xyz"
    assert_eq!(1, cj_runs.len());
    assert_eq!("\u{4e00}\u{4e8c}\u{4e09}", cj_runs[0].text);
}

// ========================================================================
// Reuse across calls
// ========================================================================

#[test]
fn segmenter_reusable_between_calls() {
    let model = latin_model();
    let mut seg = WordSegmenter::new();
    let mut all_words: Vec<String> = Vec::new();

    seg.segment("hello", &model, |buf, len, _aidx, _is_cj| {
        all_words.push(buf[..len].iter().collect());
    });
    assert_eq!(1, all_words.len());
    assert_eq!("hello", all_words[0]);

    seg.segment("world foo", &model, |buf, len, _aidx, _is_cj| {
        all_words.push(buf[..len].iter().collect());
    });
    assert_eq!(3, all_words.len());
    assert_eq!("world", all_words[1]);
    assert_eq!("foo", all_words[2]);
}

#[test]
fn extract_cj_runs_and_long_words() {
    let model = cj_model();
    let mut seg = WordSegmenter::new();
    let mut words: Vec<String> = Vec::new();
    let mut cj_runs: Vec<String> = Vec::new();

    let blob_latin = "abcdefghij";
    let blob_cj = "那只敏捷的棕色狐狸跳过了那只懒惰的狗";

    let mut text = String::new();
    text.push_str(blob_latin);
    text.push_str(blob_cj);
    // Exceed MAX_WORD_LEN (count in chars, not bytes)
    let blob_latin_chars = blob_latin.chars().count();
    let blob_cj_chars = blob_cj.chars().count();
    let mut i = 0;
    while i < MAX_WORD_LEN + 5 {
        text.push_str(blob_latin);
        i += blob_latin_chars;
    }
    // Exceed MAX_CJ_WORD_LEN
    i = 0;
    while i < MAX_CJ_WORD_LEN + 5 {
        text.push_str(blob_cj);
        i += blob_cj_chars;
    }

    seg.segment(&text, &model, |buf, len, _aidx, is_cj| {
        let s: String = buf[..len].iter().collect();
        if is_cj {
            cj_runs.push(s);
        } else {
            words.push(s);
        }
    });

    assert_eq!(2, words.len());
    assert_eq!(3, cj_runs.len());
    assert_eq!(blob_latin, words[0]);
    assert_eq!(blob_cj, cj_runs[0]);
    assert_eq!(MAX_WORD_LEN, words[1].chars().count());
    assert_eq!(MAX_CJ_WORD_LEN, cj_runs[1].chars().count());
}

// ========================================================================
// Edge cases
// ========================================================================

#[test]
fn only_spaces() {
    assert_eq!(0, segment_words("   ").len());
}

#[test]
fn only_punctuation() {
    assert_eq!(0, segment_words("---...!!!").len());
}

#[test]
fn single_char_words() {
    let words = segment_words("a b c");
    assert_eq!(3, words.len());
    assert_eq!("a", words[0].text);
    assert_eq!("b", words[1].text);
    assert_eq!("c", words[2].text);
}

#[test]
fn digit_does_not_poison_non_latin() {
    let model = cj_model();
    let (words, cj_runs) = segment_with_model("123\u{4e00}\u{4e8c}", &model);
    // CJ word should not be poisoned by digits
    let all_cj: Vec<&Word> = cj_runs.iter().chain(
        words.iter().filter(|w| w.text.contains('\u{4e00}') || w.text.contains('\u{4e8c}'))
    ).collect();
    assert!(!all_cj.is_empty(), "CJ word should be emitted despite adjacent digits");
}
